concordium-consensus/src/Concordium/GlobalState/Basic/TreeState.hs

```diff
@@ -238,38 +246,46 @@ instance (bs ~ BlockState m, BS.BlockStateStorage m, Monad m, MonadIO m, MonadSt
-                  transactionTablePurgeCounter += 1
-                  let wmdcu = WithMetadata{wmdData=cu,..}
-                  transactionTable .= (tt
-                          & (ttNonFinalizedChainUpdates . at' uty . non emptyNFCU . nfcuMap . at' sn . non Set.empty %~ Set.insert wmdcu)
-                          & (ttHashMap . at' trHash ?~ (bi, Received slot)))
-                  return (TS.Added bi)
-                else return TS.ObsoleteNonce
-          Just (_, Finalized{}) ->
-            return TS.ObsoleteNonce
+          Nothing -> do
+            -- If the transaction was not present in the `TransactionTable` then we verify it now
+            -- and adding it based on the verification result.
+            -- Only transactions which can possible be valid at the stage of execution are being added
+            -- to the `TransactionTable`.
+            -- Verifying the transaction here as opposed to `doReceiveTransactionInternal` avoids
+            -- verifying a transaction which have been both received individually and as part of a block twice.
+            verRes <- runReaderT (TS.runProtocolVersionedReaderT (TVer.verify ts bi)) verResCtx
+            if TS.definitelyNotValid verRes (verResCtx ^. TS.isTransactionFromBlock) then return $ TS.NotAdded verRes
+            else 
+              case wmdData of
+                NormalTransaction tr -> do
+                  let sender = accountAddressEmbed (transactionSender tr)
+                      nonce = transactionNonce tr
+                  if (tt ^. ttNonFinalizedTransactions . at' sender . non emptyANFT . anftNextNonce) <= nonce then do
+                    transactionTablePurgeCounter += 1
+                    let wmdtr = WithMetadata{wmdData=tr,..}
+                    transactionTable .= (tt & (ttNonFinalizedTransactions . at' sender . non emptyANFT . anftMap . at' nonce . non Map.empty %~ Map.insert wmdtr verRes)
+                                            & (ttHashMap . at' trHash ?~ (bi, Received slot verRes)))
+                    return (TS.Added bi verRes)
+                  else return TS.ObsoleteNonce
+                CredentialDeployment{} -> do
+                  transactionTable . ttHashMap . at' trHash ?= (bi, Received slot verRes)
+                  return (TS.Added bi verRes)
+                ChainUpdate cu -> do
+                  let uty = updateType (uiPayload cu)
+                      sn = updateSeqNumber (uiHeader cu)
+                  if (tt ^. ttNonFinalizedChainUpdates . at' uty . non emptyNFCU . nfcuNextSequenceNumber) <= sn then do
+                    transactionTablePurgeCounter += 1
+                    let wmdcu = WithMetadata{wmdData=cu,..}
+                    transactionTable .= (tt
+                            & (ttNonFinalizedChainUpdates . at' uty . non emptyNFCU . nfcuMap . at' sn . non Set.empty %~ Set.insert (wmdcu, verRes))
```

Comment:
Because `normal` transactions are being deleted as part of `purgeTransaction`. Having them as a tuple made it smell a bit as we should purge the transaction no matter the verification result. So using `Map.Map Transaction TVer.VerificationResult` made the code cleaner in this case. 

However we do not perform the same type of operation on `chain updates` as they are not purged 'directly' so I thought a Map was a bit overkill for this case.

https://github.com/Concordium/concordium-node/blob/main/concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs#L776

Maybe this should be revised? I've been trying the prevent touching too much of semantics with this PR.

Reply:
I fixed this to be a map now